internal/namespace/reachabilitygraph.go

```diff
@@ -0,0 +1,156 @@
+package namespace
+
+import (
+	"context"
+	"fmt"
+
+	core "github.com/authzed/spicedb/pkg/proto/core/v1"
+	"github.com/authzed/spicedb/pkg/tuple"
+)
+
+// ReachabilityGraph is a helper struct that provides an easy way to determine all entrypoints
+// for a subject of a particular type into a schema, for the purpose of walking from the subject
+// to a specific resource relation.
+type ReachabilityGraph struct {
+	ts *NamespaceTypeSystem
+}
+
+// ReachabilityEntrypoint is an entrypoint into the reachability graph for a subject of particular
+// type.
+type ReachabilityEntrypoint struct {
+	re             *core.ReachabilityEntrypoint
+	parentRelation *core.RelationReference
+}
+
+// ContainingRelationOrPermission is the relation or permission containing this entrypoint.
+func (re ReachabilityEntrypoint) ContainingRelationOrPermission() *core.RelationReference {
+	return re.parentRelation
+}
+
+// IsDirectResult returns whether the entrypoint, when evaluated, becomes a direct result of
+// the parent relation/permission. A direct result only exists if the entrypoint is not contained
+// under an intersection or exclusion, which makes the entrypoint's object merely conditionally
+// reachable.
+func (re ReachabilityEntrypoint) IsDirectResult() bool {
+	return re.re.ResultStatus == core.ReachabilityEntrypoint_DIRECT_OPERATION_RESULT
+}
+
+// ReachabilityGraphFor returns a reachability graph for the given namespace.
+func ReachabilityGraphFor(ts *ValidatedNamespaceTypeSystem) *ReachabilityGraph {
+	return &ReachabilityGraph{ts.NamespaceTypeSystem}
+}
+
+// AllEntrypointsForSubjectToResource returns the entrypoints into the reachability graph, starting
+// at the given subject type and walking to the given resource type.
+func (rg *ReachabilityGraph) AllEntrypointsForSubjectToResource(
+	ctx context.Context,
+	subjectType *core.RelationReference,
+	resourceType *core.RelationReference,
+) ([]ReachabilityEntrypoint, error) {
+	return rg.entrypointsForSubjectToResource(ctx, subjectType, resourceType, reachabilityFull)
+}
+
+// OptimizedEntrypointsForSubjectToResource returns the *optimized* set of entrypoints into the
+// reachability graph, starting at the given subject type and walking to the given resource type.
+//
+// The optimized set will skip branches on intersections and exclusions in an attempt to minimize
+// the number of entrypoints.
+func (rg *ReachabilityGraph) OptimizedEntrypointsForSubjectToResource(
+	ctx context.Context,
+	subjectType *core.RelationReference,
+	resourceType *core.RelationReference,
+) ([]ReachabilityEntrypoint, error) {
+	return rg.entrypointsForSubjectToResource(ctx, subjectType, resourceType, reachabilityOptimized)
```

Comment:
Optimized right now is only doing the pruning, it is not concerned with which branch is better. To do so will require more exploratory work in my opinion, so I wanted to do the simplest thing now until such time as we find a good solution for a stats based solution. For example, a branch may be better because its *intermediate* object sets are smaller, even if the leaf is larger, so we definitely need to have a discussion about how to pick what is "best"